util/testutil/integration/dockerd.go

```diff
@@ -0,0 +1,125 @@
+package integration
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"time"
+
+	"github.com/docker/docker/testutil/daemon"
+)
+
+const dockerdBinary = "dockerd"
+
+type logTAdapter struct {
+	Name string
+	Logs map[string]*bytes.Buffer
+}
+
+func (l logTAdapter) Logf(format string, v ...interface{}) {
+	if buf, ok := l.Logs[l.Name]; !ok || buf == nil {
+		l.Logs[l.Name] = &bytes.Buffer{}
+	}
+	fmt.Fprintf(l.Logs[l.Name], format, v...)
+}
+
+// InitDockerdWorker registers a dockerd worker with the global registry.
+func InitDockerdWorker() {
+	Register(&dockerd{})
+}
+
+type dockerd struct{}
+
+func (c dockerd) Name() string {
+	return dockerdBinary
+}
+
+func (c dockerd) New(cfg *BackendConfig) (b Backend, cl func() error, err error) {
+	if err := requireRoot(); err != nil {
+		return nil, nil, err
+	}
+
+	deferF := &multiCloser{}
+	cl = deferF.F()
+
+	defer func() {
+		if err != nil {
+			deferF.F()()
+			cl = nil
+		}
+	}()
+
+	workDir, err := ioutil.TempDir("", "integration")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cmd, err := daemon.NewDaemon(
+		workDir,
+		daemon.WithTestLogger(logTAdapter{
+			Name: "creatingDaemon",
+			Logs: cfg.Logs,
+		}),
+		daemon.WithContainerdSocket(""),
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("new daemon error: %q, %s", err, formatLogs(cfg.Logs))
+	}
+
+	err = cmd.StartWithError()
+	if err != nil {
+		return nil, nil, err
+	}
+	deferF.append(cmd.StopWithError)
+
+	logs := map[string]*bytes.Buffer{}
+	if err := waitUnix(cmd.Sock(), 5*time.Second); err != nil {
+		return nil, nil, fmt.Errorf("dockerd did not start up: %q, %s", err, formatLogs(logs))
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	deferF.append(func() error { cancel(); return nil })
+
+	dockerAPI, err := cmd.NewClient()
+	if err != nil {
+		return nil, nil, err
+	}
+	deferF.append(dockerAPI.Close)
+
+	listener, err := net.Listen("tcp", ":0")
```

Comment:
nit: this could be `unix` as well that might be bit cleaner

Reply:
Good call; didn't do this originally because I wasn't sure if we needed Windows support, but it sounds like some magic is happening and it will keep working anyways. Adding it.